Source/Charts/Data/Implementations/Standard/ChartDataSet.swift

```diff
@@ -195,58 +196,10 @@ open class ChartDataSet: ChartBaseDataSet
     /// An empty array if no Entry object at that index.
     open override func entriesForXValue(_ xValue: Double) -> [ChartDataEntry]
     {
-        var entries = [ChartDataEntry]()
-        
-        var low = startIndex
-        var high = endIndex - 1
-        
-        while low <= high
-        {
-            var m = (high + low) / 2
-            var entry = self[m]
-            
-            // if we have a match
-            if xValue == entry.x
-            {
-                while m > 0 && self[m - 1].x == xValue
-                {
-                    m -= 1
-                }
-                
-                high = endIndex
-                
-                // loop over all "equal" entries
-                while m < high
-                {
-                    entry = self[m]
-                    if entry.x == xValue
-                    {
-                        entries.append(entry)
-                    }
-                    else
-                    {
-                        break
-                    }
-                    
-                    m += 1
-                }
-                
-                break
-            }
-            else
-            {
-                if xValue > entry.x
-                {
-                    low = m + 1
-                }
-                else
-                {
-                    high = m - 1
-                }
-            }
-        }
-        
-        return entries
+        let match: (ChartDataEntry) -> Bool = { $0.x == xValue }
+        let i = partitioningIndex(where: match)
+        guard i < endIndex else { return [] }
+        return self[i...].prefix(while: match)
```

Comment:
in terms of performance, I came across this https://forums.swift.org/t/performance-of-prefix-while/12993/3

@jjatie do you think we should use `lazy` here?

Reply:
looks like it. Good find

Reply:
pushed

Reply:
@jjatie https://swiftrocks.com/lazy-sequences-in-swift-and-how-they-work go to "LazyCollection Performance Traps"

I also find this one and point out that `lazy` for Sequence and Collection have sutle difference. Could you remind me what type of `self[i...]` is here? Sequence or Collection?

Reply:
And this is why I shouldn't code review so late at night! `self[i...]` produces a `Slice<ChartDataSet>` which is a `Collection`.

The reason `lazy` performed better in the first link is because `2...` defines a `Range` with bounds from `2` to `Int.max`. Calling and algorithm on this will cause a sequence to be created with every value between those bounds. Calling `lazy` on this will only create elements of the sequence until `prefix(while:)`'s predicate fails.

In our example, all of the values are already instantiated in the data set, so calling `lazy` won't improve performance (in fact it can actually negatively effect performance). I will revert the change.

Reply:
sorry :( I was not very familiar with this topic, thank god I read to the bottom...